EC2_TAG_VOLUMES.py

```diff
@@ -0,0 +1,144 @@
+        ec2_client = client_factory.build_client("ec2")
+        resource_type = configuration_item.get("resourceType")
+        if resource_type == "AWS::EC2::Volume":
+            volume_id = configuration_item.get("resourceId")
+            if configuration_item.get("configuration").get("attachments"):
+                instance_id = configuration_item.get("configuration").get("attachments")[0].get("instanceId")
+            # Scenario:3 EBS Volume not attached to an EC2 instance.
+            if not configuration_item.get("configuration").get("attachments"):
+                return [Evaluation(ComplianceType.NOT_APPLICABLE)]
+            volume_tags = configuration_item.get("configuration").get("tags")
+            instance_tags = get_instance_tags(ec2_client, instance_id)
+            tags_to_apply = compare_instance_tags_to_volume_tags(instance_tags, volume_tags)
+            # Scenario:1 EBS Volume Tags includes same tags as EC2 Instance its attached to.
+            if not tags_to_apply:
+                return [Evaluation(ComplianceType.COMPLIANT)]
+            # Scenario:2 EBS Volume Tags does not include the same tags as EC2 Instance its attached to.
+            annotation = "EBS Volume needs to have EC2 Instance Tags applied."
+            return [Evaluation(ComplianceType.NON_COMPLIANT, annotation=annotation)]
+
+        if resource_type == "AWS::EC2::Instance":
+            instance_id = configuration_item.get("resourceId")
+            instance_tags = configuration_item.get("configuration").get("tags")
+            block_device_mappings = configuration_item.get("configuration").get("blockDeviceMappings")
+            all_volumes = get_volumes_attached_to_instance(block_device_mappings)
+            evaluations = list()
+            for volume in all_volumes:
+                volume_id = volume
+                volume_tags = get_volume_tags(ec2_client, volume_id)
+                tags_to_apply = compare_instance_tags_to_volume_tags(instance_tags, volume_tags)
+                # Scenario:1 EBS Volume Tags includes same tags as EC2 Instance its attached to.
+                if not tags_to_apply:
+                    evaluations.append(Evaluation(ComplianceType.COMPLIANT, volume_id, DEFAULT_RESOURCE_TYPE))
+                # Scenario:2 EBS Volume Tags does not include the same tags as EC2 Instance its attached to.
+                if tags_to_apply:
+                    annotation = "EBS Volume needs to have EC2 Instance Tags applied."
+                    evaluations.append(
+                        Evaluation(ComplianceType.NON_COMPLIANT, volume_id, DEFAULT_RESOURCE_TYPE, annotation)
+                    )
+            return evaluations
+
+    def evaluate_parameters(self, rule_parameters):
+        valid_rule_parameters = rule_parameters
+        return valid_rule_parameters
+
+
+def get_volume_tags(ec2_client, volume_id):
+    response = ec2_client.describe_tags(Filters=[{"Name": "resource-id", "Values": [volume_id]}])
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The API method `describe_tags` returns paginated results instead of all results. Consider using the pagination API or checking one of the following keys in the response to verify that all results were returned: `IsTruncated`, `Marker`, `NextMarker`, `NextToken`.

[Learn more](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/paginators.html)